### PR TITLE
ARTEMIS-1265 JaCoCo profile for getting code coverage report

### PR DESCRIPTION
**Pull request by dudaerich**
Added two maven profiles for:
 - generating JaCoCo exec files
 - generating JaCoCo reports

### CODE REVIEW

**Comment by dudaerich** [comment]
Example of Code coverage HTML report.

[jacoco-report.zip](https://github.com/apache/activemq-artemis/files/1115338/jacoco-report.zip)

File: docs/hacking-guide/en/code-coverage-report.md [hunk 1]
```diff
@@ -0,0 +1,37 @@
+# Code coverage report
+
+## Getting JaCoCo exec files
+
+Before you can generate code coverage report by JaCoCo tool,
+you need to get data about what lines of code were executed
```
Comment (clebertsuconic) [inline docs/hacking-guide/en/code-coverage-report.md]: You're not telling how to generate this information ^^

how to generate the data about what lines where executed?

Comment (clebertsuconic) [inline docs/hacking-guide/en/code-coverage-report.md]: You're not always generating coverage reports.. are you?

Comment (dudaerich) [inline docs/hacking-guide/en/code-coverage-report.md]: No, coverage reports are not generated always. Later in the paragraph it is written that you have to run tests with `jacoco` maven profile to generate the data. The profile is also used in the example command below the text.

File: pom.xml [hunk 2]
```diff
@@ -948,6 +966,169 @@
...
+                           <goal>merge</goal>
+                        </goals>
+                     </execution>
+                  </executions>
+                  <configuration>
+                     <fileSets>
+                        <fileSet implementation="org.apache.maven.shared.model.fileset.FileSet">
+                           <directory>${activemq.basedir}</directory>
+                           <includes>
+                              <include>**/*.exec</include>
```
Comment (jdanekrh) [inline pom.xml]: afaik, the merge+ant thing is a workaround for . latest jacoco has `report-aggregate` goal which should take care of it, http://www.eclemma.org/jacoco/trunk/doc/report-aggregate-mojo.html. The way this tends to be used is that one selects a module which does not hold any code (something with docs, or artemis-distribution, in our case, maybe) and makes it depend on all other modules (to ensure that the coverage report is written only after .exec files were generated for all modules). JaCoCo itself does it this way, https://github.com/jacoco/jacoco/blob/86856fef9959d90acdddc9d15c7b3c5f587fe086/org.jacoco.doc/pom.xml#L130.

I do not know if it is any way better or worse than this solution. Just a thought.

Comment (dudaerich) [inline pom.xml]: I looked at the `report-aggregate` goal and if I understand it correctly it doesn't fit to our case. Let me explain.

You can run this goal in the module with tests. The plugin automatically collects all jars/classes on which the module with test depends and use them for generating the report. So it generates separate report for each module with tests.

What this PR does is that at first all tests are started with Jacoco agent, so each module with tests generates jacoco.exec file. Then ant goal collects these files using the pattern `**/target/jacoco.exec` and generates report.

The `merge` goal is not necessary for generating the report. However it is useful if you want to merge all jacoco.exec files into the one and use it somewhere else. E.g. you can take the jacoco.exec file and merge it with different test suite - e.g. with wildfly's tests.

Comment (jdanekrh) [inline pom.xml]: > So it generates separate report for each module with tests.

It generates report for all module's dependencies together. So, if you run it in module that does not have any sources of its own, and it depends on all other modules in the project, there is only the dependencies report, which ends up being report for your whole project.

See for yourself:

```
$ git clone git@github.com:jacoco/jacoco.git
$ cd jacoco
$ mvn verify
$ firefox org.jacoco.doc/target/site/jacoco-aggregate/index.html
```

> Element	Missed Instructions	Cov.
> org.jacoco.examples	411563	   57%
> org.jacoco.agent.rt	187966	   83%
> jacoco-maven-plugin	1741,680   90%
> ...
> Total	1,002 of 23,389	95%

(I noticed that the Jacoco project is using some ant magic too, so it looks even using this target does not get rid of all ant.)

Comment (dudaerich) [inline pom.xml]: I understand that jars and classes can be gathered as module dependencies but how the goal locates `jacoco.exec` files which are spread in all modules with some tests?

Anyway if using of the goal requires adding of new module, I prefer to use Ant task, which doesn't require that.

Comment (jdanekrh) [inline pom.xml]: The plugin looks at all dependencies of the current module and examines them for jacoco.exec files.

New module or reusing existing one.

As I said, this was just a suggestion. Whatever works is fine with me.

**Comment by clebertsuconic** [comment]
I gave this a try and liked it a lot.. merging it 👍

**Pull request closed without merging**